pkg/platform/kube/crd/client.go

```diff
@@ -0,0 +1,379 @@
+			} else {
+				err = multierror.Append(err, fmt.Errorf("kubernetes configuration file %q", kubeconfig))
+			}
+			return "", err
+		}
+
+		// if it's an empty file, switch to in-cluster config
+		if info.Size() == 0 {
+			glog.Info("using in-cluster configuration")
+			return "", nil
+		}
+	}
+	return kubeconfig, nil
+}
+
+// CreateRESTConfig for cluster API server, pass empty config file for in-cluster
+func CreateRESTConfig(kubeconfig string) (restconfig *rest.Config, err error) {
+	if kubeconfig == "" {
+		restconfig, err = rest.InClusterConfig()
+	} else {
+		restconfig, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	}
+
+	if err != nil {
+		return
+	}
+
+	version := schema.GroupVersion{
+		Group:   config.IstioAPIGroup,
+		Version: config.IstioAPIVersion,
+	}
+
+	restconfig.GroupVersion = &version
+	restconfig.APIPath = "/apis"
+	restconfig.ContentType = runtime.ContentTypeJSON
+
+	types := runtime.NewScheme()
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			for _, kind := range knownTypes {
+				scheme.AddKnownTypes(version, kind.object, kind.collection)
+			}
+			meta_v1.AddToGroupVersion(scheme, version)
+			return nil
+		})
+	err = schemeBuilder.AddToScheme(types)
+	restconfig.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(types)}
+
+	return
+}
+
+// NewClient creates a client to Kubernetes API using a kubeconfig file.
+// Use an empty value for `kubeconfig` to use the in-cluster config.
+// If the kubeconfig file is empty, defaults to in-cluster config as well.
+func NewClient(config string, descriptor config.Descriptor) (*Client, error) {
+	for _, typ := range descriptor {
+		if _, exists := knownTypes[typ.Type]; !exists {
+			return nil, fmt.Errorf("missing known type for %q", typ.Type)
+		}
+	}
+
+	kubeconfig, err := resolveConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	restconfig, err := CreateRESTConfig(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	dynamic, err := rest.RESTClientFor(restconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	out := &Client{
+		descriptor: descriptor,
+		restconfig: restconfig,
+		dynamic:    dynamic,
+	}
+
+	return out, nil
+}
+
+// RegisterResources sends a request to create CRDs and waits for them to initialize
+func (cl *Client) RegisterResources() error {
+	clientset, err := apiextensionsclient.NewForConfig(cl.restconfig)
+	if err != nil {
+		return err
+	}
+
+	for _, schema := range cl.descriptor {
+		name := schema.Plural + "." + config.IstioAPIGroup
+		rd := &apiextensionsv1beta1.CustomResourceDefinition{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: name,
+			},
+			Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+				Group:   config.IstioAPIGroup,
+				Version: config.IstioAPIVersion,
+				Scope:   apiextensionsv1beta1.NamespaceScoped,
+				Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+					Singular: schema.Type,
```

Comment:
@ayj @kyessenov I set the Singular name specifically here (without it I ran into other naming problems), so I think '-' should just work.